### PR TITLE
[JENKINS-60563] Replace MD5 with SHA-256 in ConsistentHash

### PR DESCRIPTION
**Pull request by Wadeck**
- Also update outdated link
- Reformat some of the code
- Adjust the tests
- Ignore the performance test
- Correct typos

ℹ️ No usage of that class in the ecosystem _(Edit: outside Jenkins core that is using it for Queue)_
⚠️ For CloudBees, please check [Slack](https://cloudbees.slack.com/archives/CBWN3CULA/p1603555920429500).

See [JENKINS-60563](https://issues.jenkins-ci.org/browse/JENKINS-60563).

### Proposed changelog entries

* (internal) Modernize the hashing function of ConsistentHash

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [x] (If applicable) Jira issue is well described
- [x] Changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
  * Fill-in the `Proposed changelog entries` section only if there are breaking changes or other changes which may require extra steps from users during the upgrade
- [x] Appropriate autotests or explanation to why this change has no tests
- [n/a] For dependency updates: links to external changelogs and, if possible, full diffs

### Maintainer checklist

Before the changes are marked as `ready-for-merge`: 

- [ ] There are at least 2 approvals for the pull request and no outstanding requests for change
- [ ] Conversations in the pull request are over OR it is explicit that a reviewer does not block the change
- [ ] Changelog entries in the PR title and/or `Proposed changelog entries` are correct
- [ ] Proper changelog labels are set so that the changelog can be generated automatically
- [ ] If the change needs additional upgrade steps from users, `upgrade-guide-needed` label is set and there is a `Proposed upgrade guidelines` section in the PR title. ([example](https://github.com/jenkinsci/jenkins/pull/4387))
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins-ci.org/issues/?filter=12146)).

### CODE REVIEW

**Review by timja — APPROVED** [review]

**Comment by timja** [comment]
> No usage of that class in the ecosystem

why have it then?

**Comment by Wadeck** [comment]
> why have it then?

Used in Jenkins core only :) _(description updated)_

**Review by StefanSpieker — APPROVED** [review]
Replacing the MD5 is a good idea 👍

File: core/src/test/java/hudson/util/ConsistentHashTest.java [hunk 1]
```diff
@@ -183,18 +185,20 @@ public String hash(String str) {
      * This test doesn't fail but it's written to measure the performance of the consistent hash function with large data set.
      */
     @Test
+    @Ignore("Helper test for performance, no assertion")
```
Comment (res0nance) [inline core/src/test/java/hudson/util/ConsistentHashTest.java:188]: Fine with me, is the performance difference for this implementation noteworthy?

**Review by res0nance — APPROVED** [review]

File: core/src/test/java/hudson/util/ConsistentHashTest.java (hunk 1 above)
Comment (Wadeck) [reply core/src/test/java/hudson/util/ConsistentHashTest.java:188]: No tested due to [Hash function comparison](https://automationrhapsody.com/md5-sha-1-sha-256-sha-512-speed-performance/#:~:text=SHA%2D1%20is%20fastest%20hashing,and%2023.4%25%20for%20longer%20strings.), the difference should be around 5-10% between MD5 and SHA-256.

It's not like we are moving from MD5 to BCrypt :)

**Review by jvz — APPROVED** [review]
Seems fine, though I wonder if there's a non-cryptographic hash function that would work better for this use case.

**Review by oleg-nenashev — APPROVED** [review]
Fine with me. Maybe it is a good time to deprecate the class also.

We may merge it in 24 hours if there is no negative feedback. Please see the [merge process documentation](https://github.com/jenkinsci/jenkins/blob/master/docs/MAINTAINERS.adoc#merge-process) for more information about the merge process

**Pull request merged**